cgen: naming an array `array` results in c error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.2 5a1f3cd.4e8a2ce
**OS:** Linux version 5.10.10-arch1-1

**What did you do?**
```v
struct Container {
pub mut:
	name string
}

fn main() {
	mut array := []&Container{}

	array << &Container{ name: 'a' }

	for elem in array {
		println(elem)
	}
}
```

**What did you expect to see?**
Compile and run without error, printing the struct
 
**What did you see instead?**
```
==================
/tmp/v/bug.5101542897339626366.tmp.c: In function ‘main__main’:
/tmp/v/bug.5101542897339626366.tmp.c:8467:7: error: expected ‘;’ before ‘_t69’
 8467 |  array _t69 = array;
      |       ^~~~~
      |       ;
/tmp/v/bug.5101542897339626366.tmp.c:8468:28: error: ‘_t69’ undeclared (first use in this function)
 8468 |  for (int _t70 = 0; _t70 < _t69.len; ++_t70) {
      |                            ^~~~
/tmp/v/bug.5101542897339626366.tmp.c:8468:28: note: each undeclared identifier is reported only once for each function it appears in
```

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -1292,7 +1292,8 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 		g.expr(it.cond)
 		g.writeln(';')
 		i := if it.key_var in ['', '_'] { g.new_tmp_var() } else { it.key_var }
-		op_field := if it.cond_type.is_ptr() { '->' } else { '.' } + if it.cond_type.share() == .shared_t { 'val.' } else { '' }
+		op_field := if it.cond_type.is_ptr() { '->' } else { '.' } +
+			if it.cond_type.share() == .shared_t { 'val.' } else { '' }
 		g.writeln('for (int $i = 0; $i < $tmp${op_field}len; ++$i) {')
 		if it.val_var != '_' {
 			if val_sym.kind == .function {

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -1286,14 +1286,13 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 		g.writeln('// FOR IN array')
 		styp := g.typ(it.val_type)
 		val_sym := g.table.get_type_symbol(it.val_type)
-		cond_type_is_ptr := it.cond_type.is_ptr()
 		tmp := g.new_tmp_var()
-		tmp_type := if cond_type_is_ptr { 'array *' } else { 'array' }
-		g.write('$tmp_type $tmp = ')
+		g.write(g.typ(it.cond_type))
+		g.write(' $tmp = ')
 		g.expr(it.cond)
 		g.writeln(';')
 		i := if it.key_var in ['', '_'] { g.new_tmp_var() } else { it.key_var }
-		op_field := if cond_type_is_ptr { '->' } else { '.' }
+		op_field := if it.cond_type.is_ptr() { '->' } else { '.' }
 		g.writeln('for (int $i = 0; $i < $tmp${op_field}len; ++$i) {')
 		if it.val_var != '_' {
 			if val_sym.kind == .function {
@@ -1353,9 +1352,9 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 		g.writeln('// FOR IN map')
 		idx := g.new_tmp_var()
 		atmp := g.new_tmp_var()
-		atmp_styp := g.typ(it.cond_type)
-		arw_or_pt := if it.cond_type.nr_muls() > 0 { '->' } else { '.' }
-		g.write('$atmp_styp $atmp = ')
+		arw_or_pt := if it.cond_type.is_ptr() { '->' } else { '.' }
+		g.write(g.typ(it.cond_type))
+		g.write(' $atmp = ')
 		g.expr(it.cond)
 		g.writeln(';')
 		g.writeln('for (int $idx = 0; $idx < $atmp${arw_or_pt}key_values.len; ++$idx) {')

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -1292,7 +1292,7 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 		g.expr(it.cond)
 		g.writeln(';')
 		i := if it.key_var in ['', '_'] { g.new_tmp_var() } else { it.key_var }
-		op_field := if it.cond_type.is_ptr() { '->' } else { '.' }
+		op_field := if it.cond_type.is_ptr() { '->' } else { '.' } + if it.cond_type.share() == .shared_t { 'val.' } else { '' }
 		g.writeln('for (int $i = 0; $i < $tmp${op_field}len; ++$i) {')
 		if it.val_var != '_' {
 			if val_sym.kind == .function {

--- a/vlib/v/tests/array_test.v
+++ b/vlib/v/tests/array_test.v
@@ -9,3 +9,12 @@ fn test_for_in_array_named_array() {
 		assert *elem == 2
 	}
 }
+
+fn test_for_in_shared_array_named_array() {
+	shared array := &[1]
+	rlock array {
+		for elem in array {
+			assert elem == 1
+		}
+	}
+}

--- a/vlib/v/tests/array_test.v
+++ b/vlib/v/tests/array_test.v
@@ -1,0 +1,11 @@
+fn test_for_in_array_named_array() {
+	mut array := [1]
+	for elem in array {
+		assert elem == 1
+	}
+	for mut elem in array {
+		assert *elem == 1
+		elem = 2
+		assert *elem == 2
+	}
+}

